[Fix #18] Handle GitHub review requests

danuker opened the pull request:
This is a PR on top of #25; you should probably review that one first.

I removed almost all Trac behavior, and added the `pull_request` handler for review requests, which triggers the `_setNeedsReview` method.

Tell me when/if I should deploy it, so you can easily test it.

reviewers: @adiroiban

danuker commented on chevah/github_hooks_server/tests/test_handler.py:
```diff
@@ -794,8 +773,8 @@ def test_getTicketFromTitle(self):
-            b'Pull request has no ticket id in title: '
-            b'Simle words 12 r\xc3\x89sume.')
+            'Pull request has no ticket id in title: '
+            'Simle words 12 r\xc9sume.')
```
For convenience, I weakened this test; it only checks the string itself, no longer the byte representation.
Should I decode the string as UTF-8 bytes, and compare them?

danuker commented:
This is ready for review!
needs-review

adiroiban commented on chevah/github_hooks_server/handler.py:
```diff
@@ -56,6 +47,33 @@ def ping(self, event):
+        """
+        action = event.content.get('action')
+        if action != 'review_requested':
+            logging.warning(f"No handler for pull_request action '{action}'.")
```
I think that this should be logged as debug.
We might have many other normal actions for a PR, like "created, assigned or closed"

And maybe also include `ready_for_review` . This is when a PR is no longer a draft.

danuker replied:
`ready_for_review` would indeed be nice! No need to fumble about selecting someone to request a review from, if you already typed them in the description.

Then, they appear automatically in the "Reviewers" list after they reviewed.

adiroiban commented on chevah/github_hooks_server/handler.py:
```diff
@@ -56,6 +47,33 @@ def ping(self, event):
+
     def pull_request_review(self, event):
         """
         Called when a PR review overview message is left.
```
```suggestion
        Called when a "PR Review" action is made.
        
        See https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads#pull_request_review
        
        Things like approving or rejecting a PR via a review.
```

adiroiban commented on chevah/github_hooks_server/handler.py:
```diff
@@ -56,6 +47,33 @@ def ping(self, event):
 
+    def pull_request(self, event):
+        """
+        Called, among other cases, when a PR review is requested.
+
+        Triggers the same actions as the needs-review command to issue_comment.
+
+        The `review_requested` action is under the `pull_request` event:
+        https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads#pull_request
```
I am a bit confused. But this is due to GitHub API design and docs.
Your link is great.

I also found these 2 pages that are not that great
* https://docs.github.com/en/developers/webhooks-and-events/events/github-event-types#pullrequestevent
* https://docs.github.com/en/rest/reference/pulls


What is the difference between this method and `pull_request_review` ? 

I think that docstring should be brief. In the future it might not trigger the same actions as "needs-review".

For example, we might want to use the "draft" status to mark a PR as WIP - in work and then a draft action will trigger the same action as "needs-changes"

```suggestion
        Called for actions on a PR, exception the actions done by a reviewer.

        See: https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads#pull_request
```

danuker replied:
Indeed, the difference is that `pull_request` is for **non-review** PR changes, and `pull_request_review` is for **reviews**. Your wording captures it.

adiroiban commented on chevah/github_hooks_server/handler.py:
```diff
@@ -14,25 +14,16 @@ class Handler(object):
+        self._github = github
         if not self._github:
-            raise RuntimeError('Failed to init GitHut.')
+            raise RuntimeError('Failed to init GitHub.')
```
Just asking. Are all the exceptions logged by default?

danuker replied:
We have an all-exception handler in `hook` in `server.py`:

```python
except:
    import traceback
    logging.error(
        f'Failed to process "{event_name}":\n'
        f'{content}\n'
        f'{traceback.format_exc()}'
        )
    return func.HttpResponse(
        body="Error:003: Internal error", status_code=500
        )
```

But this particular error would be missed, since I see the Handler is created at top-level, not inside the hook method.

I will move all meaningful code inside the try block. So, the only uncaught (and unlogged) errors would be also visible offline.

Thank you for this observation!

adiroiban replied:
No problem.
There is also this option - https://stackoverflow.com/a/16993115/539264

adiroiban commented on chevah/github_hooks_server/handler.py:
```diff
@@ -41,7 +32,7 @@ def dispatch(self, event):
         if handler is None:
             message = f'No handler for "{event.name}"'
-            logging.error(message)
+            logging.warning(message)
```
I think that this should be debug.

it's OK to not have handlers for a certain action.

adiroiban commented on chevah/github_hooks_server/handler.py:
```diff
@@ -56,6 +47,33 @@ def ping(self, event):
+        pull_id = event.content['pull_request']['number']
+        reviewers = self._getReviewers(event.content['pull_request']['body'])
+
+        logging.info(
```
this is ok...but in the future maybe we can also pass the event to _setNeedsReview and have a single log.

I see that _setNeedsReview does it's own extra debug logging.

danuker replied:
I'll keep it, since we call _setNeedsReview from two places now, and it is useful to know who called it.

adiroiban reviewed (changes requested):
All good. Thanks.

Only a minor comment about refactoring to allow future PR actions like drafts.

We are not using drafts and we might never use drafts.
I hope that we can make this tools generic enough to be used by other projects. I know that other projects are using a draft as a way to say that "WIP. this PR is not ready. don't bother to review or look at this code yet."

danuker commented:
Would like it if you took another look.

I cleaned up, and deleted everything mentioning the ticket ID, since we don't use it anymore (at least for now). Only the PR ID matters.

needs-review

danuker commented:
I redeployed. I am testing it. needs-changes

adiroiban reviewed (approved):
Just a test... I will re-request  the review

adiroiban reviewed (approved):
Looks good. Thanks.

Can be merged.


For the future we should see how to handle the "assignee" part...but I think that is ok how it is now.

If you are the "owner" of the PR and the ticket is still waiting for reviewers, you should not be "assignee" as the ball is not in your fields.
